includes/Modules/Sign_In_With_Google.php

```diff
@@ -82,23 +140,53 @@ public function handle_google_auth() {
 			exit;
 		}
 
-		$client_id = $this->get_settings()->get()['clientID'];
-		$id_token  = $this->context->input()->filter( INPUT_POST, 'credential' );
+		$id_token = $this->context->input()->filter( INPUT_POST, 'credential' );
 		try {
-			$client  = new Google_Client( array( 'client_id' => $client_id ) );
-			$payload = $client->verifyIdToken( $id_token );
+			$payload = $this->client->verifyIdToken( $id_token );
+		} catch ( \Exception $e ) {
+			wp_safe_redirect( add_query_arg( 'error', 'google_auth_invalid_request', wp_login_url() ) );
+			exit;
+		}
+
+		if ( is_null( $payload ) || empty( $payload ) || ! array_key_exists( 'sub', $payload ) || empty( $payload['sub'] ) || ! array_key_exists( 'email', $payload ) || empty( $payload['email'] ) ) {
+			wp_safe_redirect( add_query_arg( 'error', 'google_auth_invalid_request', wp_login_url() ) );
+			exit;
+		}
 
-			if ( empty( $payload ) ) {
-				wp_safe_redirect( add_query_arg( 'error', 'google_auth_invalid_request', wp_login_url() ) );
-				exit;
-			}
+		$google_user_id    = sanitize_text_field( $payload['sub'] );
+		$google_user_email = sanitize_email( $payload['email'] );
 
-			// @TODO implement further flow using $payload in #9339.
+		// Check if there are any existing WordPress users connected to this Google account.
+		// The user ID is used as the unique identifier because users can change the email on their Google account.
+		$existing_users = get_users(
+			array(
+				// phpcs:ignore WordPress.DB.SlowDBQuery.slow_db_query_meta_key
+				'meta_key'   => self::SIGN_IN_WITH_GOOGLE_USER_ID_OPTION,
+				// phpcs:ignore WordPress.DB.SlowDBQuery.slow_db_query_meta_value
+				'meta_value' => $google_user_id,
+				'number'     => 1,
+			)
+		);
+		if ( ! empty( $existing_users ) && $existing_users[0] instanceof WP_User ) {
+			return $this->login_user( $existing_users[0] );
+		}
 
-		} catch ( \Exception $e ) {
-			wp_safe_redirect( add_query_arg( 'error', 'google_auth_invalid_request', wp_login_url() ) );
+		// Find an existing user that matches the email and link to their Google account by store their user ID in user meta.
+		$existing_user = get_user_by( 'email', $google_user_email );
+		if ( $existing_user instanceof WP_User ) {
```

Comment:
The `get_user_by` function returns either a user instance or `false`.
```suggestion
		if ( $existing_user ) {
```